### PR TITLE
Ft authentication protect routes160877586

### PR DESCRIPTION
**Pull request by MuhweziDeo**
**What does this PR do?**
Reviews user authentication for all endpoints

**Description of Task to be completed?**

- creation a JSON web token for authentication
- Ensure all endpoints for a normal user and admin user are protected
- Normal users should not perform admin functionality

* How should this be manually tested?**

-  `git` clone https://github.com/MuhweziDeo/Fast-foods-Fast-challenge3.git
- cd into cloned repository
- run git checkout ft-authentication-protect-routes160877586 to switch to branch
- create a virtual environment using virtualenv venv
- activate environment using source venv/bin/activate
- install dependencies using pip -r install requirements.txt
- run py.test --cov

**What are the relevant pivotal tracker stories?**
#160877586
**Screenshots**
screenshot from 2018-10-03 16-11-23
screenshot from 2018-10-03 16-12-50

### CODE REVIEW

**Review by Kasulejoseph — COMMENTED** [review]
your work is good..

**Comment by kafuuma** [comment]
I have reviewed your code 
A few points to note
It's not a good programming practice to use multiple if-else statements,  It also affects you when you are writing test cases for those multiple else statements
I advise you to refactor your  code

**Comment by Sebuliba-Adrian** [comment]
@MuhweziDeo Have you addressed @kafuuma's feedback?

**Comment by Sebuliba-Adrian** [comment]
I do not know what the hashes are doing in you PR. Please remove them.

**Comment by Sebuliba-Adrian** [comment]
Any reason you branched off ```ft-delete-meal-option-160877559 from ft-authentication-protect-routes160877586``` instead of develop?

**Comment by mubstimor** [comment]
Good work, remember to keep database credentials out of version control

**Comment by MuhweziDeo** [comment]
> I have reviewed your code
> A few points to note
> It's not a good programming practice to use multiple if-else statements, It also affects you when you are writing test cases for those multiple else statements
> I advise you to refactor your code

thanks am going to refactor code and element then

**Comment by MuhweziDeo** [comment]
> Any reason you branched off `ft-delete-meal-option-160877559 from ft-authentication-protect-routes160877586`?

i thought it bolds them.. I have fixed it

**Comment by MuhweziDeo** [comment]
> @MuhweziDeo Have you addressed @kafuuma's feedback?

currently going to work on it
thanks for the feedback

**Comment by MuhweziDeo** [comment]
> Any reason you branched off `ft-delete-meal-option-160877559 from ft-authentication-protect-routes160877586`?

it was the last feature i implemented before adding authentication

**Comment by MuhweziDeo** [comment]
> Good work, remember to keep database credentials out of version control

yeah will use enviroment variables

File: api/routes/views.py [hunk 1]
```diff
@@ -1,12 +1,14 @@
...
 db = DB(user='postgres', password='sudo',
         dbname=' ', host='localhost')
 
 api = Api(app, prefix='/api/v2', version='2.0', title='Fast-Foods-Api')
-
+jwt = JWTManager(app)
+app.config['JWT_SECRET_KEY'] = 'thisissecretkhahxcahiahiac'
 
 db.create_db_tables()
 # db.drop_all_tables('orders','users','fastfoods')
```
Comment (Sebuliba-Adrian) [inline api/routes/views.py]: @MuhweziDeo I thought we talked about leaving commented code in the code base. Whats this now?

File: api/routes/views.py [hunk 2]
```diff
@@ -1,12 +1,14 @@
...
 from api.app import app
 from api.models.db import DB
 
 db = DB(user='postgres', password='sudo',
         dbname=' ', host='localhost')
 
 api = Api(app, prefix='/api/v2', version='2.0', title='Fast-Foods-Api')
-
+jwt = JWTManager(app)
+app.config['JWT_SECRET_KEY'] = 'thisissecretkhahxcahiahiac'
```
Comment (Sebuliba-Adrian) [inline api/routes/views.py:11]: Do not expose your secret keys in the code. keep them in a ```.env``` file that you can use to access all you environment variables . In fact keep all sensitive info in this file.

File: api/models/db.py [hunk 3]
```diff
@@ -96,7 +100,22 @@ def register_user(self, username, password):
         except(Exception, psycopg2.DatabaseError) as error:
             print(error)
             # return None
-            return {'message': 'Please choose a another name because username {} already exits'.format(username)}
+            return {'message': 'Unable to create account'}
+
+    def register_admin(self, username, password):
+        try:
+            password = self.hash_password(password)
+            # new_user=User(username,password)
```
Comment (Sebuliba-Adrian) [inline api/models/db.py:108]: Commented code?

**Review by Sebuliba-Adrian — CHANGES_REQUESTED** [review]
@MuhweziDeo .Separate your test files into chunks ,``` login test file alone```,  ```registration test file alone```, ```orders CRUD``` and ```menu CRUD``` all alone.  For separation of concerns.

File: api/routes/views.py (hunk 2 above)
Comment (MuhweziDeo) [reply api/routes/views.py:11]: thanks working on it in deployment

File: api/models/db.py (hunk 3 above)
Comment (MuhweziDeo) [reply api/models/db.py:108]: fixed it thanks

**Pull request closed without merging**